File: Sensor Testing/ADC/mcp3424.py
# Copyright (C) 2023 Dawson J. Gullickson All rights reserved.

from typing import Union

from enum import Enum

import busio # type: ignore

# ---== THE MCP3424 DATASHEET ==---
# https://www.mouser.com/datasheet/2/268/22088c-3136953.pdf
# Section 5.2: Configuration Register
# Section 5.3 I2C Serial Communications
# Figure 6.1: I2C Connection Diagram

FSR = 2.048
"""Full Scale Range"""

class SampleRate(Enum):
    """Available Sample Rates"""
    _240 = 0
    _60 = 1
    _15 = 2
    _3_75 = 3
    
BITS_PER_RATE = {
    SampleRate._240: 12,
    SampleRate._60: 14,
    SampleRate._15: 16,
    SampleRate._3_75: 18
}
"""Data Rates Per Sampling Rate"""

ConfigTuple = tuple[int, bool, bool, SampleRate, int]

def construct_config_byte(
    channel: int,
    ready: bool,
    continuous_mode: bool,
    sample_rate: SampleRate,
    gain: int
) -> int:
    """Constructs a configuration byte.
    
    :param bool ready:
    :param int channel: Channel to select, in the range [1,4]
    :param bool continuous_mode:
    :param SampleRate sample_rate:
    :param int gain: log2 of the desired gain, in the range [0,3].
    """
    if channel < 1 or 4 < channel:
        raise ValueError("Channel must be an integer in the range [1,4].")
    if gain < 0 or 3 < gain:
        raise ValueError("Gain must be an integer in the range [0,3].")
    
    config = ready << 7
    config += (channel - 1) << 5
    config += continuous_mode << 4
    config += sample_rate.value << 2
    config += gain
    return config

def parse_config_byte(config: int) -> ConfigTuple:
    """Parses a configuration byte."""
    channel = ((config >> 5) % 4) + 1
    ready = config >> 7 == 1
    continuous_mode = (config >> 4) % 2 == 1
    sample_rate = SampleRate((config >> 2) % 4)
    gain = config % 4

    return channel, ready, continuous_mode, sample_rate, gain

def update_config_byte(
    config: int,
    channel: Union[int, None] = None,
    ready: Union[bool, None] = None,
    continuous_mode: Union[bool, None] = None,
    sample_rate: Union[SampleRate, None] = None,
    gain: Union[int, None] = None
) -> int:
    """Updates a configuration byte."""
    (
        channel_old,
        ready_old,
        continuous_mode_old,
        sample_rate_old,
        gain_old
    ) = parse_config_byte(config)

    channel = channel if channel is not None else channel_old
    ready = ready if ready is not None else ready_old
    continuous_mode = continuous_mode if continuous_mode is not None else continuous_mode_old
    sample_rate = sample_rate if sample_rate is not None else sample_rate_old
    gain = gain if gain is not None else gain_old
    
    config_updated = construct_config_byte(
        channel, ready, continuous_mode, sample_rate, gain
    )
    return config_updated

def construct_address_byte(address: int) -> int:
    """Constructs an MCP3424 address byte."""
    if address < 0 or 7 < address:
        raise ValueError("Address must be an integer in the range [0,7].")
    
    addr = 0b1101 << 3 # Set by manufacturer
    addr += address
    return addr

class ADC:
    bus: busio.I2C
    address: int
    config: int

    def __init__(
        self,
        bus: busio.I2C,
        address: int,
        channel: int = 1,
        ready: bool = False,
        continuous_mode: bool = True,
        sample_rate: SampleRate = SampleRate._240,
        gain: int = 0
    ) -> None:
        self.bus = bus
        self.address = address
        self.config = construct_config_byte(
            channel, ready, continuous_mode, sample_rate, gain
        )

    def update_config(
        self,
        channel: Union[int, None] = None,
        ready: Union[bool, None] = None,
        continuous_mode: Union[bool, None] = None,
        sample_rate: Union[SampleRate, None] = None,
        gain: Union[int, None] = None
    ) -> None:
        self.config = update_config_byte(
            self.config, channel, ready, continuous_mode, sample_rate, gain
        )

    def configure(self) -> None:
        addr = construct_address_byte(self.address)
        
        try:
            assert self.bus.try_lock()
            self.bus.write_byte(addr, self.config)
        except AssertionError as e: raise e
        finally: self.bus.unlock()

    def update_and_configure(
        self,
        channel: Union[int, None] = None,
        ready: Union[bool, None] = None,
        continuous_mode: Union[bool, None] = None,
        sample_rate: Union[SampleRate, None] = None,
        gain: Union[int, None] = None
    ) -> None:
        self.update_config(
            channel, ready, continuous_mode, sample_rate, gain
        )
        self.configure()
        
    def read(self) -> float:
        addr = construct_address_byte(self.address)

        # Read 3 bytes from ADC
        buffer = bytearray(3)
        try:
            assert self.bus.try_lock()
            self.bus.readfrom_into(addr, buffer)
        except AssertionError as e: raise e
        finally: self.bus.unlock()
        data = int.from_bytes(buffer, byteorder="big")

        # Convert to voltage
        _, _, _, sample_rate, gain = parse_config_byte(self.config)
        bits = BITS_PER_RATE[sample_rate]
        converted = (data / (1 << bits)) * (FSR / gain)

        return converted

File: Sensor Testing/ADC/mcp3424.c
// Copyright 2024 by Dawson J. Gullickson

#include "mcp3424.h"

#include <fcntl.h>
#include <stdint.h>
#include <unistd.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>

const double FULL_SCALE_RANGE = 2.048;

uint8_t bits_per_rate(SampleRate sample_rate) {
    switch (sample_rate) {
    case HZ240: return 12;
    case HZ60: return 14;
    case HZ15: return 16;
    case HZ3_75: return 18;
    }
    return 0; //NOTE Unreachable
}

uint8_t construct_config_byte(Config config) {
    uint8_t byte = 0;
    byte |= ((config.ready) ? 1:0) << 7;
    byte |= config.channel << 5;
    byte |= ((config.continuous_mode) ? 1:0) << 4;
    byte |= config.sample_rate << 2;
    byte |= config.gain;
    return byte;
}

Config parse_config_byte(uint8_t byte) {
    Config config;
    config.ready = (byte >> 7) == 1;
    config.channel = ((byte >> 5) % 4);
    config.continuous_mode = (byte >> 4) % 2 == 1;
    config.sample_rate = (byte >> 2) % 4;
    config.gain = byte % 4;
    return config;
}


uint8_t construct_address_byte(uint8_t address) {
    uint8_t byte = 0b1101 << 3; // Set by manufacturer
    byte |= address;
    return byte;
}

int _ADC_set_slave_addr(ADC adc) {
    uint8_t addr = construct_address_byte(adc.address);
    return ioctl(adc.bus, I2C_SLAVE, addr);
}

int ADC_init(ADC *adc, const char *i2c_bus, uint8_t address) {
    // Open bus
    int bus = open(i2c_bus, O_RDWR);
    if (bus < 0) return bus; // Error
    
    // Initialize
    adc->bus = bus;
    adc->address = address;

    return 0;
}

int ADC_configure(ADC adc) {
    uint8_t config = construct_config_byte(adc.config);
    uint8_t buffer[1] = { config };
    if (_ADC_set_slave_addr(adc) < 0) return -1;
    if (write(adc.bus, buffer, 1) != 1) return -2;
    return 0;
}

int ADC_read(ADC adc, double *value) {
    // Read 3 bytes from ADC
    uint8_t buffer[3];
    if (_ADC_set_slave_addr(adc) < 0) return -1;
    if (read(adc.bus, buffer, 3) != 3) return -2;

    // Get bits
    uint8_t bits;
    if (adc.config.continuous_mode)
        bits = bits_per_rate(adc.config.sample_rate);
    else
        bits = 18;
    
    // Combine bytes
    uint32_t data = buffer[2];
    data |= ((uint32_t)buffer[1]) << 8;
    data |= ((uint32_t)buffer[0]) << 16;
    if (bits == 18) data <<= 8;

    // Convert to voltage
    uint8_t gain = 1 << adc.config.gain;
    int32_t data_signed = ~data + 1;
    double LSB = FULL_SCALE_RANGE / (1 << (uint32_t)(bits - 1));
    *value = (double)data_signed * (LSB / gain);

    return 0;
}